staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting/validation.go

```diff
@@ -0,0 +1,121 @@
+}
+
+type validator struct {
+	rootSchema *structuralschema.Structural
+}
+
+// validate is the recursive step func for the validation. insideMeta is true if s specifies TypeMeta or ObjectMeta.
+func (v *validator) validate(pth *field.Path, s *structuralschema.Structural, f SurroundingObjectFunc, insideMeta bool) (field.ErrorList, error) {
+	if s == nil {
+		return nil, nil
+	}
+
+	if s.XEmbeddedResource {
+		insideMeta = false
+		f = NewRootObjectFunc().WithTypeMeta(metav1.TypeMeta{APIVersion: "validation/v1", Kind: "Validation"})
+		v2 := *v
+		v = &v2
+		v.rootSchema = s
+	}
+
+	allErrs := field.ErrorList{}
+
+	if s.Default.Object != nil {
+		validator := validate.NewSchemaValidator(s.ToGoOpenAPI(), nil, "", strfmt.Default)
+
+		if insideMeta {
+			obj, _, err := f(runtime.DeepCopyJSONValue(s.Default.Object))
+			if err != nil {
+				return nil, fmt.Errorf("failed to prune default value: %v", err)
+			}
+
+			// check ObjectMeta/TypeMeta and everything else
+			if err := schemaobjectmeta.Coerce(nil, obj, v.rootSchema, true, false); err != nil {
+				allErrs = append(allErrs, field.Invalid(pth.Child("default"), s.Default.Object, fmt.Sprintf("must result in valid metadata: %v", err)))
+			} else if errs := schemaobjectmeta.Validate(nil, obj, v.rootSchema, true); len(errs) > 0 {
+				allErrs = append(allErrs, field.Invalid(pth.Child("default"), s.Default.Object, fmt.Sprintf("must result in valid metadata: %v", errs.ToAggregate())))
+			} else if errs := apiservervalidation.ValidateCustomResource(pth.Child("default"), s.Default.Object, validator); len(errs) > 0 {
+				allErrs = append(allErrs, errs...)
+			}
+		} else {
+			// check whether default is pruned
+			pruned := runtime.DeepCopyJSONValue(s.Default.Object)
+			pruning.Prune(pruned, s, s.XEmbeddedResource)
+			if !reflect.DeepEqual(pruned, s.Default.Object) {
+				allErrs = append(allErrs, field.Invalid(pth.Child("default"), s.Default.Object, "must not have unknown fields"))
+			}
+
+			// check ObjectMeta/TypeMeta and everything else
+			if err := schemaobjectmeta.Coerce(pth.Child("default"), s.Default.Object, s, s.XEmbeddedResource, false); err != nil {
+				allErrs = append(allErrs, err)
+			} else if errs := schemaobjectmeta.Validate(pth.Child("default"), s.Default.Object, s, s.XEmbeddedResource); len(errs) > 0 {
+				allErrs = append(allErrs, errs...)
+			} else if errs := apiservervalidation.ValidateCustomResource(pth.Child("default"), s.Default.Object, validator); len(errs) > 0 {
+				allErrs = append(allErrs, errs...)
+			}
+		}
+	}
+
+	// do not follow additionalProperties because defaults are forbidden there
+
+	if s.Items != nil {
+		if errs, err := v.validate(pth.Child("items"), s.Items, f.Index(), insideMeta); err != nil {
+			return nil, err
+		} else {
+			allErrs = append(allErrs, errs...)
+		}
+	}
+
+	for k, subSchema := range s.Properties {
+		subInsideMeta := insideMeta
+		if s.XEmbeddedResource && (k == "metadata" || k == "apiVersion" || k == "kind") {
```

Comment:
trying to remember, is it not possible to define defaults inside root metadata?

Reply:
No, we don't allow that (because it might lead to non-standard behaviour or even incompatibility with api machinery).